vm/runtime.ts

```diff
@@ -1349,130 +1352,153 @@ module J2ME {
-      var updateGlobalObject = true;
-      if (nativeMethod) {
-        linkWriter && linkWriter.writeLn("Method: " + methodDescription + " -> Native / Override");
-        fn = nativeMethod;
-        methodType = MethodType.Native;
-        methodInfo.state = MethodState.Compiled;
-      } else {
-        fn = findCompiledMethod(klass, methodInfo);
-        if (fn) {
-          linkWriter && linkWriter.greenLn("Method: " + methodDescription + " -> Compiled");
-          methodType = MethodType.Compiled;
-          // Save method info so that we can figure out where we are bailing
-          // out from.
-          jitMethodInfos[fn.name] = methodInfo;
-          updateGlobalObject = false;
-          methodInfo.state = MethodState.Compiled;
-        } else {
-          linkWriter && linkWriter.warnLn("Method: " + methodDescription + " -> Interpreter");
-          methodType = MethodType.Interpreted;
-          fn = prepareInterpretedMethod(methodInfo);
+      // var key = methodType !== MethodType.Interpreted ? MethodType[methodType] : methodInfo.implKey;
+      // var key = MethodType[methodType] + " " + methodInfo.implKey;
+      nativeCounter.count(key);
+      var s = bytecodeCount;
+      try {
+        methodTimeline.enter(key);
+        var r;
+        switch (arguments.length) {
+          case 0:
+            r = fn.call(this);
+            break;
+          case 1:
+            r = fn.call(this, a);
+            break;
+          case 2:
+            r = fn.call(this, a, b);
+            break;
+          case 3:
+            r = fn.call(this, a, b, c);
+            break;
+          default:
+            r = fn.apply(this, arguments);
         }
+        methodTimeline.leave(key, s !== bytecodeCount ? { bytecodeCount: bytecodeCount - s } : undefined);
+      } catch (e) {
+        methodTimeline.leave(key, s !== bytecodeCount ? { bytecodeCount: bytecodeCount - s } : undefined);
+        throw e;
       }
-      if (false && methodTimeline) {
-        fn = profilingWrapper(fn, methodInfo, methodType);
-        updateGlobalObject = true;
-      }
+      return r;
+    };
+  }
 
-      if (traceWriter) {
-        fn = tracingWrapper(fn, methodInfo, methodType);
-        updateGlobalObject = true;
-      }
+  function tracingWrapper(fn: Function, methodInfo: MethodInfo, methodType: MethodType) {
+    return function() {
+      var args = Array.prototype.slice.apply(arguments);
+      traceWriter.enter("> " + MethodType[methodType][0] + " " + methodInfo.implKey + " " + (methodInfo.stats.callCount ++));
+      var s = performance.now();
+      var value = fn.apply(this, args);
+      traceWriter.outdent();
+      return value;
+    };
+  }
 
-      methodInfo.fn = fn;
+  export function tryLinkMethod(methodInfo: MethodInfo) {
+    if (methodInfo.fn) {
+      return methodInfo.fn;
+    }
+    linkKlassMethod(methodInfo.classInfo.klass, methodInfo);
+    assert (methodInfo.fn);
+    return methodInfo.fn;
```

Comment:
Nit: *try* in the name suggests this could fail. But no code accounts for that failure. Do we really expect this function to fail? If not, we should call it simply *linkMethod* (and make the assertion conditional on *!release*). Otherwise we need a better story for handling runtime failures of this function in release builds.